### DISCUSSION TITLE
Reading Energy History from Vitotronic WO1C Vitocal 200-S

### DISCUSSION THREAD

**Discussion by philippoo66** [philippoo66/optolink-splitter#9] [General]
Der Ursprung dieses Threads liegt unter open3e und geht ca. [hier](https://github.com/open3e/open3e/discussions/94#discussioncomment-9084617) los. Der Lösung nahe kommen wir etwa [hier](https://github.com/open3e/open3e/discussions/94#discussioncomment-9598339). Ab jetzt führen wir das Thema in diesem Thread weiter. :-)

**Comment by timbrd** [comment]
Wie noch im alten Thread geschrieben, vermute ich, dass die Heizung nur Tageswerte der vergangenen 7 Tage gibt. Ist für mich aber kein Problem, da ich die Tageswerte sowieso in Homeassistant schiebe und in dessen Datenbank vorhalte.

**Reply by philippoo66** [reply]
> 30 Sekunden müssen nicht sein. Aber eine feine Auflösung über den Tag hinweg in Homeassistant zu haben ist schon gut. So kann ich erkennen, wann und in welchen Zeiträumen die Heizung besonders viel Strom gezogen hat oder es größere Sprünge in der Effizienz gegeben hat.

in welchem Intervall denkst du?

wird das denn wirklich laufend aktualisiert? aber für die vergangenen Tage ändert sich ja nix mehr, oder?

**Reply by philippoo66** [reply]
wenn wir nur den aktiven Tag nehmen, können wir das natürlich schon in die Polling Liste mit rein schreiben. 1 Telegramm mehr macht da auch nix...

**Reply by timbrd** [reply]
> > 30 Sekunden müssen nicht sein. Aber eine feine Auflösung über den Tag hinweg in Homeassistant zu haben ist schon gut. So kann ich erkennen, wann und in welchen Zeiträumen die Heizung besonders viel Strom gezogen hat oder es größere Sprünge in der Effizienz gegeben hat.
> 
> in welchem Intervall denkst du?
> 
> wird das denn wirklich laufend aktualisiert? aber für die vergangenen Tage ändert sich ja nix mehr, oder?

Ein Großteil meiner Geräte, die ich über Homeassistant verwalte, arbeiten aus Gründen der Energiesparsamkeit nach dem Push-Prinzip und melden nur Änderungen. Es gibt eine ganze Reihe von Nachteilen an der Methode. Bspw. passiert es immer wieder, dass ein Gerät ein Problem hat und das dann nicht auffällt.

Diejenigen, die aktiv abgefragt werden, haben in der Regel ein Intervall von 5 Minuten. Hier mal ein Beispiel des Energieverbrauchs einer Steckdose, an dem mein Gefrierschrank hängt:

<img width="1201" alt="Steckdose" src="https://github.com/philippoo66/optolink-splitter/assets/7645199/d9df1b74-7806-401b-b630-183bf712a9e0">

Und nein, die vergangenen Tage sind natürlich fix.

**Reply by timbrd** [reply]
> wenn wir nur den aktiven Tag nehmen, können wir das natürlich schon in die Polling Liste mit rein schreiben. 1 Telegramm mehr macht da auch nix...

Es wäre schon gut, wenn man auch die Historie der letzten 7 Tage erhalten würde. Falls bspw. mal Homeassistant oder der Pi über einen Tag hinweg ausfällt (Urlaub etc.), dann wäre es toll, wenn man die letzten Datenpunkte eines jeden Tages nachgetragen bekäme.
Ich muss mir darüber aber nochmal Gedanken machen.

**Reply by philippoo66** [reply]
das Ding passt schlecht in's Schema des DataPoint Lesens wenn wir die vier Werte explizit übergeben wollen. Ich denke, wir hängen am Ende der Polling Loop eine special function dafür an. da können wir ggf auch einen Zähler rein machen, dass sie nur jedes x-te Mal tatsächlich ausgeführt wird.

Wie soll das denn aussehen auf MQTT? Das Datum / der Zeitstempel wird ja eigentlich in der InfluxDB (oder welche nutzt du?) eh mit notiert. als Tag immer "0" anzugeben macht auch keinen Sinn. also einfach 4 'leaves' mit Bezeichnung und Wert wie bei den anderen Datenpunkten, nur eben jedes mal 4? Wenn ja - wie sollen die benannt werden?

**Reply by philippoo66** [reply]
> Es wäre schon gut, wenn man auch die Historie der letzten 7 Tage erhalten würde. 

sorry, das hat sich überschnitten. geht natürlich auch. Sag einfach, wie du dir das vorstellst auf MQTT

**Reply by philippoo66** [reply]
> vermute ich, dass die Heizung nur Tageswerte der vergangenen 7 Tage gibt.

hast du eigentlich mal versucht, mehr als 6 Tage zurück zu lesen? das byte würde ja 0..255 'fassen'...

**Reply by philippoo66** [reply]
ich hab das mit den Werten / Byte-Indexen übrigens auch grade gebogen. is im Repo.

**Reply by timbrd** [reply]
> ich hab das mit den Werten / Byte-Indexen übrigens auch grade gebogen. is im Repo.

Danke, die Werte werden jetzt auch richtig ausgegeben.

**Reply by timbrd** [reply]
> > vermute ich, dass die Heizung nur Tageswerte der vergangenen 7 Tage gibt.
> 
> hast du eigentlich mal versucht, mehr als 6 Tage zurück zu lesen? das byte würde ja 0..255 'fassen'...

Ja, habe ich. Aber ab Tag 7 werden keine Werte mehr geliefert. Stattdessen erhalte ich eine Fehlermeldung (Index out of Range oder so ähnlich)

**Reply by timbrd** [reply]
> > Es wäre schon gut, wenn man auch die Historie der letzten 7 Tage erhalten würde.
> 
> sorry, das hat sich überschnitten. geht natürlich auch. Sag einfach, wie du dir das vorstellst auf MQTT

Ich spiele das erste Mal mit einem MQTT und Homeassistant herum und habe jetzt erst mal etwas recherchiert, welche Möglichkeiten ich habe. Tatsächlich kann man keine Werte für vergangene Eregnisse setzen, alle eintreffenden Daten werden mit einem Timestamp versehen, die Homeassistant als Zeitpunkt "der Erzeugung" interpretiert. Um historische Daten zu importieren müsste ich eine Integration (Plugin) schreiben und selbst dann gibt es eine Reihe von Hürden.
Damit fällt mir gerade keine sinnvolle Verwendung für die Energiewerte der letzten 6 Tage in Homeassistant ein.
Ich bräuchte daher nur die Daten des aktuellen Tages. Diese können gerne in einem kurzen Intervall abgefragt werden bspw. 60 Sekunden.

**Reply by timbrd** [reply]
> das Ding passt schlecht in's Schema des DataPoint Lesens wenn wir die vier Werte explizit übergeben wollen. Ich denke, wir hängen am Ende der Polling Loop eine special function dafür an. da können wir ggf auch einen Zähler rein machen, dass sie nur jedes x-te Mal tatsächlich ausgeführt wird.
> 
> Wie soll das denn aussehen auf MQTT? Das Datum / der Zeitstempel wird ja eigentlich in der InfluxDB (oder welche nutzt du?) eh mit notiert. als Tag immer "0" anzugeben macht auch keinen Sinn. also einfach 4 'leaves' mit Bezeichnung und Wert wie bei den anderen Datenpunkten, nur eben jedes mal 4? Wenn ja - wie sollen die benannt werden?

Homeassistant kann mit JSON umgehen. Wenn ich die Werte der 7 Tage benötigt hätte, dann wäre JSON mein Wunschformat gewesen, mit dem wir genau das Problem umgangen hätten. :)

**Reply by philippoo66** [reply]
> Ich bräuchte daher nur die Daten des aktuellen Tages. Diese können gerne in einem kurzen Intervall abgefragt werden bspw. 60 Sekunden.

oki, mach ich. wie soll es denn dann aussehen (du weisst schon, der MQTT tree / die leaves )

![grafik](https://github.com/philippoo66/optolink-splitter/assets/122479122/0ceea966-ba5a-4c4a-83e8-f932c5b28db9)

oder anders? Mit dem Main Topic drin wär glaubich gut, weil sonst braucht es glaubich ne andre publish_mqtt()... was waren noch mal die 'sprechenden' Bezeichnungen der Daten 0..3?

**Reply by timbrd** [reply]
> > Ich bräuchte daher nur die Daten des aktuellen Tages. Diese können gerne in einem kurzen Intervall abgefragt werden bspw. 60 Sekunden.
> 
> oki, mach ich. wie soll es denn dann aussehen (du weisst schon, der MQTT tree / die leaves )

Ich würde zunächst bei der flat Struktur bleiben und später mal darüber nachdenken, ob man nicht alle Daten aus der settings_ini in einer Baumstruktur anordnen können sollte. Für die vier Werte ist der Aufwand für das Refactoring aber wahrscheinlich nicht gerechtfertigt.

> oder anders? Mit dem Main Topic drin wär glaubich gut, weil sonst braucht es glaubich ne andre publish_mqtt()... was waren noch mal die 'sprechenden' Bezeichnungen der Daten 0..3?

Ich bin ja für englische Bezeichner, momentan verwende ich:
- energy_heating_thermal
- energy_heating_electric
- energy_water_thermal
- energy_water_electric

**Reply by philippoo66** [reply]
ok, mach ich fertig und sag dann Bescheid.

**Reply by philippoo66** [reply]
ok, ist in https://github.com/philippoo66/optolink-splitter/tree/wo1c_energy

es wurde nur geändert
- optolinkvs2_switch.py
```
                        #!!! wo1c energy special !!!
                        wo1c_energy.read_energy(serViDev)
                        #!!! +++++++++++++++++++ !!!
```
Z.230 eingefügt (und der Modulimport)


es wurde hinzugefügt
- wo1c_energy.py

Es ist eine quick&dirty Implementierung. Es wird nicht geprüft, ob MQTT überhaupt angebunden und so, Adresse, Tage_zurück, Leave Names etc sind nur im Code editierbar (aber das bekämest du bestimmt hin ;-)

Das read_energy() wird jetzt am Ende jedes Polling List Zyklus aufgerufen. In dem Abschnitt

```
    # every x-th time...
    global interval
    interval += 1
    if(interval < 2):
        return
    interval = 0
```

kannst du regulieren, jedes wievielte Mal es dann tatsächlich ausgeführt wird. So eben jedes zweite Mal. Wenn du 30 sec Polling Intervall eingestellt hast, passiert es eben alle 60 Sekunden.

Probier's mal und sag Bescheid! Getestet habe ich es jetzt nicht weil ich keine wo1c habe ;-)

**Reply by philippoo66** [reply]
ps. 
> ob man nicht alle Daten aus der settings_ini in einer Baumstruktur anordnen können sollte. 

das müsste eigentlich schon gehen, wenn du statt `"WertName"` einfach `"MySubTopic/WertName"` angibst.

**Reply by timbrd** [reply]
> ok, ist in https://github.com/philippoo66/optolink-splitter/tree/wo1c_energy
> [...]
> Probier's mal und sag Bescheid! Getestet habe ich es jetzt nicht weil ich keine wo1c habe ;-)

Ich habe das gestern mal getestet und mich gewundert, dass die Werte falsch sind. Heute passen sie aber wieder.
Mir ist dann aufgefallen, dass ich nicht erwähnt habe, dass der heutige Tag nicht durch die erste zurückgelieferte Reihe abgebildet wird. Der heutige Tag ist die Reihe mit dem Index des "day of week".
Hier mal ein Beispiel für alle 7 Zahlenreihen:

<img width="688" alt="Viessmann" src="https://github.com/philippoo66/optolink-splitter/assets/7645199/d20954ce-1c92-4502-83b9-83741ab6ea33">

In diesem Beispiel muss am heutigen Montag die erste Reihe zurückgegeben werden, morgen aber die zweite, am Mittwoch die dritte usw.

**Reply by philippoo66** [reply]
achso, kein Problem. ist eingebaut. passt es jetzt?

**Reply by timbrd** [reply]
Fast. Du multipzierst alle Werte mit 10 (Zeile 52ff). Richtig wäre es, die Werte durch 10 zu dividieren.

**Reply by philippoo66** [reply]
oh, die funktion hb ich glaubich irgendwann mal geändert... passt es jetzt?

**Reply by timbrd** [reply]
Perfekt, danke!

**Reply by timbrd** [reply]
Ich lasse das jetzt mal ein paar Tage mitlaufen und überlegen mir, wie ich die ganzen Daten in Homeassistant auswerte.
